arrow-flight/src/encode.rs

```diff
@@ -0,0 +1,503 @@
+/// # use arrow_array::{ArrayRef, RecordBatch, UInt32Array};
+/// # async fn f() {
+/// # let c1 = UInt32Array::from(vec![1, 2, 3, 4, 5, 6]);
+/// # let record_batch = RecordBatch::try_from_iter(vec![
+/// #      ("a", Arc::new(c1) as ArrayRef)
+/// #   ])
+/// #   .expect("cannot create record batch");
+/// use arrow_flight::encode::FlightDataEncoderBuilder;
+///
+/// // Get an input stream of Result<RecordBatch, FlightError>
+/// let input_stream = futures::stream::iter(vec![Ok(record_batch)]);
+///
+/// // Build a stream of `Result<FlightData>` (e.g. to return for do_get)
+/// let flight_data_stream = FlightDataEncoderBuilder::new()
+///  .build(input_stream);
+///
+/// // Create a tonic `Response` that can be returned from a Flight server
+/// let response = tonic::Response::new(flight_data_stream);
+/// # }
+/// ```
+///
+/// [`FlightService::do_get`]: crate::flight_service_server::FlightService::do_get
+#[derive(Debug)]
+pub struct FlightDataEncoderBuilder {
+    /// The maximum message size (see details on [`Self::with_max_message_size`]).
+    max_batch_size: usize,
+    /// Ipc writer options
+    options: IpcWriteOptions,
+    /// Metadata to add to the schema message
+    app_metadata: Bytes,
+}
+
+/// Default target size for record batches to send.
+///
+/// Note this value would normally be 4MB, but the size calculation is
+/// somehwhat inexact, so we set it to 2MB.
+pub const GRPC_TARGET_MAX_BATCH_SIZE: usize = 2097152;
+
+impl Default for FlightDataEncoderBuilder {
+    fn default() -> Self {
+        Self {
+            max_batch_size: GRPC_TARGET_MAX_BATCH_SIZE,
+            options: IpcWriteOptions::default(),
+            app_metadata: Bytes::new(),
+        }
+    }
+}
+
+impl FlightDataEncoderBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Set the (approximate) maximum encoded [`RecordBatch`] size to
+    /// limit the gRPC message size. Defaults fo 2MB.
+    ///
+    /// The encoder splits up [`RecordBatch`]s (preserving order) to
+    /// limit individual messages to approximately this size. The size
+    /// is approximate because there additional encoding overhead on
+    /// top of the underlying data itself.
+    ///
+    pub fn with_max_message_size(mut self, max_batch_size: usize) -> Self {
+        self.max_batch_size = max_batch_size;
+        self
+    }
+
+    /// Specfy application specific metadata included in the
+    /// [`FlightData::app_metadata`] field of the the first Schema
+    /// message
+    pub fn with_metadata(mut self, app_metadata: Bytes) -> Self {
+        self.app_metadata = app_metadata;
+        self
+    }
+
+    /// Set the [`IpcWriteOptions`] used to encode the [`RecordBatch`]es for transport.
+    pub fn with_options(mut self, options: IpcWriteOptions) -> Self {
+        self.options = options;
+        self
+    }
+
+    /// Return a [`Stream`](futures::Stream) of [`FlightData`],
+    /// consuming self. More details on [`FlightDataEncoderBuilder`]
```

Comment:
Do you probably mean?

```suggestion
    /// Return a [`Stream`](futures::Stream) of [`FlightData`],
    /// consuming self. More details on [`FlightDataEncoder`]
```